networkx/algorithms/tests/test_planarity.py

```diff
@@ -436,7 +438,9 @@ def test_unsuccessful_face_traversal(self):
     @staticmethod
     def get_star_embedding(n):
         embedding = nx.PlanarEmbedding()
+        ref = None
         for i in range(1, n):
-            embedding.add_half_edge_first(0, i)
-            embedding.add_half_edge_first(i, 0)
+            embedding.add_half_edge(0, i, cw=ref)
```

Comment:
I don't think I understand the older version of this test.  Why is it OK to add a "first" edge to node 0 multiple times? Is it because "first edge" actually means "last edge"? Hence the use elsewhere of `next(reversed(self._succ[]))`...

This new test is certainly more clear.

Is there a reason to use the var `ref` here instead of `i - 1`?

Reply:
The `add_half_edge_first()` is actually adding edges to the left of the leftmost out-edge of start_node, while also working if start_node has no out-edges to begin with. The newly added edge is marked as the "first_nbr" of start_node. This is something used in the Left-Right Planarity test implemented in `LRPlanarity.lr_planarity()`. The use of "first" in the method name also confused me. To make things more cryptic, `add_half_edge_ccw()` will update the "first_nbr" if the new half-edge is added ccw to it.

I used the last position of the succs dict to keep track or "first_nbr" just because it is easier to re-insert the first_nbr at the end than it is to insert a new first_nbr at the beggining.

Reply:
The use of `ref` is for passing None in the first call (no neighbors to reference then).